### PR TITLE
Add CODEOWNERS

### PR DESCRIPTION
**Pull request by kratman**
# Description

## Change
Code owners are now specified in `.github/CODEOWNERS` and can be configured as required reviewers. This can help make it easier for maintainer-trainees to help with the project maintenance. The code owners can be required to approve pull requests, but the trainees can be granted write permission (merge approved PRs, trigger workflow runs, etc). This will make the trainees more involved in the process.

## Side effects
All members of the pybamm-team/maintainers group will be requested to review on new PRs. This can be avoided by making PRs drafts. The review requests happen only when a PR is marked as ready for review.

## Next steps
In order for this to work properly some settings need to be adjusted by a maintainer.

1) The maintainer-trainees group should have permissions upgraded to `write` access. This should let us merge approved PRs and trigger workflows.

2) The following four options should be enabled in the repo after this PR is merged. These settings will prevent non-code owners from merging code that is not approved by a code owner. Dismissing stale reviews prevents changes being pushed post-review.
image

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Review by agriyakhetarpal — APPROVED** [review]
Looks good to me, thanks @kratman! I would like to note that I am not sure how I feel about an entire team getting tagged on new PRs. In my experience, there is a lot of noise in my GitHub notifications despite the fact that I watch repositories selectively... is there a workaround? I have seen other repositories and teams work with CODEOWNERS without automatic review requests: they show up in a tooltip instead with the words "`@user` requested a review from a team as a code owner" – does that require GitHub Enterprise?

**Comment by kratman** [comment]
@agriyakhetarpal I have only worked on a single project that used CODEOWNERS. I was not admin for that repo, so I do not know how configuring notifications works with it. I know a bunch of people who setup custom email filters to handle github notifications.

I put my notification/watch settings pretty high for projects I work on, but I know it annoys some people.

**Review by Saransh-cpp — COMMENTED** [review]
Thanks for working on this, @kratman! I'll streamline this in the upcoming days to avoid sending notifications to everyone on every PR. I'll probably push to this branch itself.

**Comment by kratman** [comment]
Seems like that is a lot of fine grained control that needs to be edited and maintained over time. I did the one line version to keep it simple.

Do we really need to restrict certain modules to specific users rather than just the team as a whole? I would hope all maintainers can be trusted with the entire codebase.

**Comment by kratman** [comment]
> Edit: I cannot add maintainer trainees to the file as they do not have write access to pybamm. I'll look into the workflow permissions.

I don't think they should be in there. The original purpose was to let the trainees get more access while still restricting what they can do with the code. Once a basic code owners file is in place the trainees can safely have their rights elevated.

File: .github/CODEOWNERS
Comment (kratman) [inline .github/CODEOWNERS:35]: I feel like this is a little too complicated and it is something that we would have to keep up-to-date.

I have not played around with overlapping rules before, so take this with a grain of salt. If we use scripts as an example:
```
# Installation and other scripts
/scripts/ @martinjrobins @Saransh-cpp
```
Does this mean only Saransh and Martin could review those files or does it let any maintainer approve and only they get notifications? If it is the former, then I think that would be a bit troublesome. The docs for code owners suggests that Saransh and Martin would be the owners for that directory, which sounds like the former case.

Comment (kratman) [reply .github/CODEOWNERS:35]: Docs reference:
```
# These owners will be the default owners for everything in
# the repo. Unless a later match takes precedence,
# @global-owner1 and @global-owner2 will be requested for
# review when someone opens a pull request.
*       @global-owner1 @global-owner2

# Order is important; the last matching pattern takes the most
# precedence. When someone opens a pull request that only
# modifies JS files, only @js-owner and not the global
# owner(s) will be requested for a review.
*.js    @js-owner #This is an inline comment.
```

Comment (kratman) [reply .github/CODEOWNERS:35]: I guess in my initial change I was suggesting that we turn on the require review from code owners in order to let trainees get more access. With this change as it is now, we get more notifications, but trainees still don't get more access.

Comment (kratman) [reply .github/CODEOWNERS:35]: I think if we are not going to do the branch protection rules change, then we probably don't need to bother with a CODEOWNERS file.

Comment (brosaplanella) [reply .github/CODEOWNERS:35]: > Does this mean only Saransh and Martin could review those files or does it let any maintainer approve and only they get notifications? If it is the former, then I think that would be a bit troublesome. The docs for code owners suggests that Saransh and Martin would be the owners for that directory, which sounds like the former case.

I agree, we should aim for the latter.

For the parts of the code owned by pybamm-team/maintainers, does it mean that everyone will be requested to review?

Comment (kratman) [reply .github/CODEOWNERS:35]: A single "maintainers" group will appear in the review section and each maintainer will get a single email. I believe that after somebody reviews the maintainers fer unassigned and the reviewer is the only one tagged. It has been a while since I used one of these files, but I believe that is how it works

Comment (agriyakhetarpal) [reply .github/CODEOWNERS:35]: Does it auto-request a review or does it display an option for requesting reviews from the maintainers for those with the Triage level of permissions? It would be helpful to control the request behaviour.

Also, should we try testing this in someone's fork, since that would provide more clarity?

**Comment by kratman** [comment]
> * For pybamm-team/maintainers, the whole team would be requested for review. For instance, this PR has also requested a review from pybamm-team/maintainers.

It seems like half the maintainers would get notified with this option as well. That is sort of why I was pushing for the simpler solution.

File: .github/CODEOWNERS
Comment (kratman) [inline .github/CODEOWNERS:17]: Since rules lower down the list override ones above, wouldn't all maintainers get assigned for all the folders above? Maybe the /pybamm/* should go up higher?

**Comment by tinosulzer** [comment]
Unrelated to the general discussion here, but should we move `meshes`, `geometry`, `spatial_methods`, and `discretization` under one folder?

File: .github/CODEOWNERS
Comment (Saransh-cpp) [reply .github/CODEOWNERS:17]: `/pybamm/*` will not recursively check every subdirectory, but will only check the files directly under `pybamm/` -

```
# The `docs/*` pattern will match files like
# `docs/getting-started.md` but not further nested files like
# `docs/build-app/troubleshooting.md`.
docs/*  docs@example.com
```

Comment (kratman) [reply .github/CODEOWNERS:17]: Works for me then

File: .github/CODEOWNERS
Comment (Saransh-cpp) [inline .github/CODEOWNERS:1]: Add @agriyakhetarpal, @kratman, and @arjxn-py in this file before merging.

File: .github/CODEOWNERS
Comment (Saransh-cpp) [inline .github/CODEOWNERS:16]:
```suggestion
```

File: .github/CODEOWNERS
Comment (agriyakhetarpal) [reply .github/CODEOWNERS:1]: Thanks, can I push changes here directly?

Comment (agriyakhetarpal) [reply .github/CODEOWNERS:1]: Yes, thanks – can I push changes here directly?

Comment (agriyakhetarpal) [reply .github/CODEOWNERS:1]: Please do, I think I can't push changes here from my fork currently

Comment (kratman) [reply .github/CODEOWNERS:1]: @agriyakhetarpal I will make you a collaborator on my Fork. I have no concerns about giving you access

Comment (agriyakhetarpal) [reply .github/CODEOWNERS:1]: Thanks a lot, received and accepted!

Comment (kratman) [reply .github/CODEOWNERS:1]: @arjxn-py I don't know you as well, so I won't add you to my fork yet, but send me what you want in Slack and I will add it to the file

File: .github/CODEOWNERS
Comment (Saransh-cpp) [inline .github/CODEOWNERS:2]:
```suggestion
```
Can either remove this or arrange the usernames alphabetically

**Review by Saransh-cpp — APPROVED** [review]
Thanks for working on this @kratman! I think we have the final structure ready now. There will be a few changes once discretisation-related files are clubbed, so maybe we can merge this after #3641.

File: .github/CODEOWNERS
Comment (Saransh-cpp) [reply .github/CODEOWNERS:27]: Ah, I did this because I could not add @agriyakhetarpal to the file back then. I think the best person to maintain or own these folders would be @agriyakhetarpal, if he is comfortable doing so.

Comment (kratman) [reply .github/CODEOWNERS:27]: You can add me. I am not bothered by notifications

Comment (arjxn-py) [reply .github/CODEOWNERS:27]: You can also consider me adding for `/docs/` atleast. Nor am I bothered by notifications 🙂

Comment (agriyakhetarpal) [reply .github/CODEOWNERS:27]: Me three. I am not bothered by notifications either :) I would be comfortable maintaining both `docs/` and `examples/`.

File: .github/CODEOWNERS
Comment (agriyakhetarpal) [inline .github/CODEOWNERS:12]:
```suggestion
/pybamm/plotting/ @martinjrobins @rtimms @Saransh-cpp @tinosulzer @rtimms  @kratman @agriyakhetarpal 
```
I am happy to learn and maintain the plotting functionality too as we go along

File: .github/CODEOWNERS
Comment (kratman) [reply .github/CODEOWNERS:27]:
```suggestion
/docs/ @kratman @arjxn-py @agriyakhetarpal 

# Example scripts
/examples/ @kratman @agriyakhetarpal 
```

**Comment by kratman** [comment]
@Saransh-cpp Since #3641 is stalled, can we merge this so it can be tested and update it later?

File: .github/CODEOWNERS
Comment (Saransh-cpp) [inline .github/CODEOWNERS:27]: I'll be happy to review PRs here too!
```suggestion
/docs/ @kratman @arjxn-py @agriyakhetarpal @Saransh-cpp

# Example scripts
/examples/ @kratman @agriyakhetarpal @Saransh-cpp
```

**Comment by Saransh-cpp** [comment]
> can we merge this so it can be tested and update it later?

Let's wait for a couple of days for the remaining approvals from SC.

**Comment by kratman** [comment]
> > can we merge this so it can be tested and update it later?
> 
> Let's wait for a couple of days for the remaining approvals from SC.

@tinosulzer @brosaplanella @martinjrobins

**Pull request merged**
